File: stock-market-service/src/main/java/rs/edu/raf/strategy/OrderProcessStrategyFactory.java
package rs.edu.raf.strategy;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;
import rs.edu.raf.model.order.*;


@Component
public class OrderProcessStrategyFactory {
    @Autowired
    @Qualifier("marketOrder")
    private OrderProcessStrategy marketOrderProcessStrategy;

    @Autowired
    @Qualifier("stopOrder")
    private OrderProcessStrategy stopOrderProcessStrategy;

    @Autowired
    @Qualifier("stopLimitOrder")
    private OrderProcessStrategy stopLimitOrderProcessStrategy;

    @Autowired
    @Qualifier("allOrNoneOrder")
    private OrderProcessStrategy allOrNoneOrderProcessStrategy;

    @Autowired
    @Qualifier("limitOrder")
    private OrderProcessStrategy limitOrderProcessStrategy;

    public OrderProcessStrategy getExecutionStrategy(Order order) {
        if(order instanceof MarketOrder) {
            return marketOrderProcessStrategy;
        }

        if(order instanceof StopOrder) {
            return stopOrderProcessStrategy;
        }

        if(order instanceof StopLimitOrder) {
            return stopLimitOrderProcessStrategy;
        }

        if(order instanceof AllOrNoneOrder) {
            return allOrNoneOrderProcessStrategy;
        }

        if(order instanceof LimitOrder) {
            return limitOrderProcessStrategy;
        }

        throw new RuntimeException();
    }
}

File: stock-market-service-kotlin/src/main/kotlin/org/example/stockmarketservicekotlin/strategy/OrderProcessStrategyFactory.kt
package org.example.stockmarketservicekotlin.strategy

import org.example.stockmarketservicekotlin.model.order.*
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Component

@Component
class OrderProcessStrategyFactory(
    @Qualifier("marketOrder")
    private val marketOrderProcessStrategy: OrderProcessStrategy,


    @Qualifier("stopOrder")
    private val stopOrderProcessStrategy: OrderProcessStrategy,


    @Qualifier("stopLimitOrder")
    private val stopLimitOrderProcessStrategy: OrderProcessStrategy,


    @Qualifier("allOrNoneOrder")
    private val allOrNoneOrderProcessStrategy: OrderProcessStrategy,


    @Qualifier("limitOrder")
    private val limitOrderProcessStrategy: OrderProcessStrategy,
) {
    fun getExecutionStrategy(order: Order): OrderProcessStrategy {
        if (order is MarkerOrder) {
            return marketOrderProcessStrategy
        }

        if (order is StopOrder) {
            return stopOrderProcessStrategy
        }

        if (order is StopLimitOrder) {
            return stopLimitOrderProcessStrategy
        }

        if (order is AllOrNoneOrder) {
            return allOrNoneOrderProcessStrategy
        }

        if (order is LimitOrder) {
            return limitOrderProcessStrategy
        }

        throw RuntimeException()
    }
}